edap.py

```diff
@@ -213,14 +42,181 @@ def __init__(self, hostname, admin_cn, password, domain=None):
+
+    def create_group_dict(self, name):
+        dic = dict(
+            cn=name.encode("ASCII"), objectclass=(b"posixGroup", b"top"), gidNumber=b"500",
+        )
+        return dic
+
+    def create_group_from_dict(self, fqdn, dic):
+        modlist = ldap.modlist.addModlist(dic)
+        self.ldap.add_s(fqdn, modlist)
+
+    def create_service_group(self, name):
+        if not self.subobject_exists_at(self.ldap.SERVICES, "organizationalUnit"):
+            self.create_org_unit(self.ldap.SERVICES, self.ldap.SERVICES_GROUP)
+        if not self.subobject_exists_at(f"cn={name},{self.ldap.SERVICES}", "posixGroup"):
+            dic = self.create_group_dict(f"{name}")
+            self.create_group_from_dict(f"cn={name},{self.ldap.SERVICES_GROUP}", dic)
+
+
+class LdapUserMixin(object):
+
+    def add_user(self, uid, name, surname, password):
+        if self.subobject_exists_at("ou=people", "organizationalUnit") == 0:
+            raise ConstraintError(f"The people group '{self.ldap.PEOPLE_GROUP}' doesn't exist.")
+        if self.user_of_uid_exists(uid) > 0:
+            raise ConstraintError(f"User of uid '{uid}' already exists.")
+        modlist = self.mk_add_user_modlist(uid, name, surname, password)
+        self.ldap.add_s(f"uid={uid},{self.ldap.PEOPLE_GROUP}", modlist)
+
+    def mk_add_user_modlist(self, uid, name, surname, password):
+        mail = f"{uid}@example.com".encode("ASCII")
+        dic = dict(
+            uid=uid.encode("ASCII"), givenName=name.encode("UTF-8"),
+            mail=mail, objectclass=(b"inetOrgPerson", b"top"),
+            sn=surname.encode("UTF-8"), userPassword=_hashPassword(password),
+            cn=f"{name} {surname}".encode("UTF-8"),
+        )
+        modlist = ldap.modlist.addModlist(dic)
+        return modlist
+
+    def user_of_uid_exists(self, uid):
+        if self.subobject_exists_at("ou=people", "organizationalUnit") == 0:
+            raise ConstraintError(f"The people group '{self.ldap.PEOPLE_GROUP}' doesn't exist.")
+        found = self.ldap.search_s(f"{self.ldap.PEOPLE_GROUP}", ldap.SCOPE_ONELEVEL, f"(uid={uid})")
+        return len(found)
+
+    def uid_is_member_of_group(self, group_fqdn, uid):
+        search = f"memberUid={uid}"
+        found = self.ldap.search_s(group_fqdn, ldap.SCOPE_BASE, f"({search})")
+        return len(found)
+
+    def make_uid_member_of(self, uid, group_fqdn):
+        if self.object_exists_at(group_fqdn, "posixGroup") == 0:
+            raise ConstraintError(f"Group {group_fqdn} doesn't exist.")
+        if self.user_of_uid_exists(uid) == 0:
+            msg = f"User of uid '{uid}' doesn't exist, so we can't add it to any group."
+            raise ConstraintError(msg)
+        if self.uid_is_member_of_group(group_fqdn, uid):
+            return
+        modlist = [(ldap.MOD_ADD, "memberUid", [uid.encode("ASCII")])]
+        self.ldap.modify_s(group_fqdn, modlist)
+
+    def make_uid_member_of_division(self, uid, name):
```

Comment:
Note for later - the class should not be aware of concept of division - it makes more sense to have a method `make_uid_member_of` that would accept an object that would know what is it's name and position in the LDAP tree.

Reply:
Agree, thanks, will do it as part of #2